cow_instrument/InstrumentTree.cpp

```diff
@@ -76,44 +90,51 @@ V3D InstrumentTree::samplePos() const {
   return V3D{0, 0, 10}; // HACK!
 }
 
-size_t InstrumentTree::nDetectors() const { return m_detectorVec.size(); }
+size_t InstrumentTree::nDetectors() const { return m_detectorVec->size(); }
 
 std::unique_ptr<const InstrumentTree>
 InstrumentTree::modify(size_t node, const Command &command) const {
 
-  auto newNodes(m_nodes);
-  std::for_each(newNodes.begin(), newNodes.end(),
+  CowPtr<std::vector<Node>> newNodes(m_nodes);
+  std::for_each(newNodes->begin(), newNodes->end(),
                 [](Node &node) { node.incrementVersion(); });
 
   if (command.isMetaDataCommand()) {
     // No cascading behaviour.
-    auto &currentNode = newNodes[node];
+    Node &currentNode = newNodes->operator[](node);
     currentNode.doModify(command);
+
   } else {
     // Cascading behaviour
-      std::vector<size_t> toModify = {node};
+    std::vector<size_t> toModify = {node};
     for (size_t index = 0; index < toModify.size(); ++index) {
-      auto &currentNode = newNodes[toModify[index]];
+      Node &currentNode = newNodes->operator[](toModify[index]);
       currentNode.doModify(command);
```

Comment:
It's worth a try. Depends on what the overhead of all the checks is like, but if we can avoid unnecessary rebuilds of that Detector* vector, we should try.